### PR TITLE
OpenVINO Hybrid CPUs support internal interface

### PR DESCRIPTION
**Pull request by ivankochin**
This patch implements the interface which allows binding the work inside the arena to certain core types while working on systems with Hybrid CPUs. To implement such cross-platform (Windows and Linux) functionality the tbbbind_2_4 and HWLOC 2.4 static libraries are used.

The internal interface (namespace `custom`) is the same as the corresponding oneTBB Hybrid CPUs support API and also covers the NUMA support API to hide compile-time branching and avoid mixing incompatible HWLOC versions: 
- [oneTBB Hybrid CPUs support - constraints part](https://software.intel.com/content/www/us/en/develop/documentation/onetbb-documentation/top/onetbb-api-reference/task-arena-constraints-extensions.html)
- [oneTBB Hybrid CPUs support - info namespace part](https://software.intel.com/content/www/us/en/develop/documentation/onetbb-documentation/top/onetbb-api-reference/oneapi-tbb-info-namespace-extensions.html)
- [oneTBB NUMA support API specification - info namespace part](https://spec.oneapi.com/versions/latest/elements/oneTBB/source/info_namespace.html)
- [oneTBB NUMA support API specification - constraints part](https://spec.oneapi.com/versions/latest/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.html#_CPPv411constraints)

Now the `custom` namespace implements the following compile-time logic which chooses what library should be used for binging the work inside `tbb::task_arena` to the core type/NUMA node:

| TBB version / tbbbind_2_4 availability                      | available   | unavailable                      |
|-------------------------------------------------------------|-------------|----------------------------------|
| TBB_INTERFACE_VERSION < 11100 (no NUMA or Hybrid CPUs APIs) | tbbbind_2_4 | binding is unavailable           |
| TBB_INTERFACE_VERSION >= 11100 (includes NUMA API)          | tbbbind_2_4 | tbb (only NUMA API is available) |
| TBB_INTERFACE_VERSION >= 12020 (includes Hybrid CPUs API)   | tbb         | tbb                              |

At this patch, only the primitive logic regards choosing what binding type should be used is implemented. Advanced logic is implemented in the [separate patch](https://github.com/openvinotoolkit/openvino/pull/5261/files) based on these changes.

### CODE REVIEW

File: inference-engine/include/ie_parallel_custom_arena.hpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+//
+
+/**
+ * @brief Contains declarations and custom threading interfaces based on TBB info and task_arena APIs.
+ *
+ * @file ie_parallel_custom.hpp
+ */
+
+#pragma once
+
```
Comment (ilya-lavrenov) [inline inference-engine/include/ie_parallel_custom_arena.hpp:12]: hope this will not be a part of public interface in a final solution..

Comment (ivankochin) [reply inference-engine/include/ie_parallel_custom_arena.hpp]: At this moment it is already moved to the internal part.

File: inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+void destroy_binding_observer(binding_observer* observer) {
+    observer->observe(false);
+    delete observer;
+}
+#endif /*USE_TBBBIND_2_4*/
+
+#if TBB_NUMA_SUPPORT_PRESENT
+tbb::task_arena::constraints convert_constraints(task_arena::constraints& c) {
```
Comment (ivankochin) [inline inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp:138]:
```suggestion
tbb::task_arena::constraints convert_constraints(custom::task_arena::constraints& c) {
```

File: inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+{}
+
+void task_arena::initialize() {
+#if USE_TBBBIND_2_4
+    std::call_once(my_initialization_state, [this] {
+        tbb::task_arena::initialize();
+        my_binding_observer = detail::construct_binding_observer(
+            *this, tbb::task_arena::max_concurrency(), my_constraints);
+    });
+#elif TBB_NUMA_SUPPORT_PRESENT && TBB_HYBRID_CPUS_SUPPORT_PRESENT
```
Comment (ivankochin) [inline inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp:179]:
```suggestion
#elif TBB_NUMA_SUPPORT_PRESENT || TBB_HYBRID_CPUS_SUPPORT_PRESENT
```

File: inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+void task_arena::initialize(int max_concurrency_, unsigned reserved_for_masters) {
+#if USE_TBBBIND_2_4
+    std::call_once(my_initialization_state, [this, &max_concurrency_, &reserved_for_masters] {
+        tbb::task_arena::initialize(max_concurrency_, reserved_for_masters);
+        my_binding_observer = detail::construct_binding_observer(
+            *this, tbb::task_arena::max_concurrency(), my_constraints);
+    });
+#elif TBB_NUMA_SUPPORT_PRESENT && TBB_HYBRID_CPUS_SUPPORT_PRESENT
```
Comment (ivankochin) [inline inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp:193]:
```suggestion
#elif TBB_NUMA_SUPPORT_PRESENT || TBB_HYBRID_CPUS_SUPPORT_PRESENT
```

File: inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+void task_arena::initialize(constraints constraints_, unsigned reserved_for_masters) {
+    std::call_once(my_initialization_state, [this, &constraints_, &reserved_for_masters] {
+        my_constraints = constraints_;
+#if USE_TBBBIND_2_4
+        tbb::task_arena::initialize(info::default_concurrency(constraints_), reserved_for_masters);
+        my_binding_observer = detail::construct_binding_observer(
+            *this, tbb::task_arena::max_concurrency(), my_constraints);
+#elif TBB_NUMA_SUPPORT_PRESENT && TBB_HYBRID_CPUS_SUPPORT_PRESENT
```
Comment (ivankochin) [inline inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp:208]:
```suggestion
#elif TBB_NUMA_SUPPORT_PRESENT || TBB_HYBRID_CPUS_SUPPORT_PRESENT
```

File: inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+std::vector<numa_node_id> numa_nodes() {
+#if USE_TBBBIND_2_4
+    detail::initialize_system_topology();
+    std::vector<numa_node_id> node_indexes(detail::numa_nodes_count);
+    std::memcpy(node_indexes.data(), detail::numa_nodes_indexes, detail::numa_nodes_count * sizeof(int));
+    return node_indexes;
+#elif TBB_NUMA_SUPPORT_PRESENT
+    return tbb::info::numa_nodes();
+#else
+    return {-1};
```
Comment (ivankochin) [inline inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp:239]:
```suggestion
    return {tbb::task_arena::automatic};
```

File: inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+std::vector<core_type_id> core_types() {
+#if USE_TBBBIND_2_4
+    detail::initialize_system_topology();
+    std::vector<numa_node_id> core_type_indexes(detail::core_types_count);
+    std::memcpy(core_type_indexes.data(), detail::core_types_indexes, detail::core_types_count * sizeof(int));
+    return core_type_indexes;
+#elif TBB_HYBRID_CPUS_SUPPORT_PRESENT
+    return tbb::info::core_types();
+#else
+    return {-1};
```
Comment (ivankochin) [inline inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp:252]:
```suggestion
    return {tbb::task_arena::automatic};
```

File: inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+#else
+    return true;
+#endif /* _WIN32 && !_WIN64 */
+}
+
+int  numa_nodes_count = 0;
+int* numa_nodes_indexes = nullptr;
+
+int  core_types_count = 0;
+int* core_types_indexes = nullptr;
```
Comment (ivankochin) [inline inference-engine/src/inference_engine/threading/ie_parallel_custom_arena.cpp:78]:
```suggestion
static int  numa_nodes_count = 0;
static int* numa_nodes_indexes = nullptr;

static int  core_types_count = 0;
static int* core_types_indexes = nullptr;
```

**Review by ilya-lavrenov — APPROVED** [review]

**Comment by myshevts** [comment]
pls check that compilation succeeds with oneTBB, and old TBB (before NUMA), when either is provided via TBB_DIR

**Comment by ivankochin** [comment]
> pls check that compilation succeeds with oneTBB, and old TBB (before NUMA), when either is provided via TBB_DIR (-DTBB_DIR=)

This patch was checked on Linux with three different TBB version: 
- oneTBB 2021.2 (NUMA + Hybrid CPUs support)
- TBB 2020.3 (NUMA support only)
- TBB 2019

**Comment by myshevts** [comment]
I've measured the perf on ICX and compared to the master commit that Ivan had based his branch on. There is no perf diff for the conventional code-path.
(will need to check on the production ADL, but this also requires [my branch](https://github.com/openvinotoolkit/openvino/pull/5261/) with the hybrid logic to be merged as well) .

So I'm approving the request and will merge that tommorow (unless any showstopper comments from @apankratovantonp) so that it will be measured/tested as a part of the coming weekly.

**Review by myshevts — APPROVED** [review]

**Review by apankratovantonp — APPROVED** [review]

**Pull request merged**